File: 4.1-calculos-matematicos/Prueba-3/Variacion-2/javascript/main.js
function calculatePiGaussLegendre(digits) {
    const precision = BigInt(digits + 5);
    let a = BigInt(1);
    let b = BigInt(Math.floor(Math.sqrt(2) * 10 ** digits));
    let t = BigInt(1) * BigInt(10 ** digits) / BigInt(4);
    let p = BigInt(1);
    let numIter = Math.floor(Math.log10(digits));

    for (let i = 0; i < numIter; i++) {
        let aNext = (a + b) / BigInt(2);
        b = BigInt(Math.floor(Math.sqrt(Number(a * b)) * 10 ** digits));
        t -= p * (a - aNext) * (a - aNext);
        a = aNext;
        p *= BigInt(2);
    }

    let pi = ((a + b) * (a + b)) / (BigInt(4) * t);
    return pi;
}

async function n_digits_pi(repetitions, digits) {
    let totalTime = 0;
    let totalMemory = 0;

    const startTotal = performance.now();

    for (let i = 0; i < repetitions; i++) {
        const start = performance.now();

        const piValue = calculatePiGaussLegendre(digits);

        const end = performance.now();
        const memoryUsage = performance.memory ? (performance.memory.usedJSHeapSize / (1024 * 1024)) : 0;

        totalTime += (end - start);
        totalMemory += memoryUsage;
    }

    const endTotal = performance.now();
    const totalExecTime = ((endTotal - startTotal) / 1000).toFixed(2);

    const avgTime = Number((totalTime / repetitions).toFixed(2));
    const avgMemory = Number((totalMemory / repetitions).toFixed(2));

    const outputDiv = document.getElementById("javascript-output");

    let timeDiv = document.createElement("div");
    timeDiv.textContent = `Total ET: ${totalExecTime} s`;
    outputDiv.appendChild(timeDiv);

    let avgTimeDiv = document.createElement("div");
    avgTimeDiv.textContent = `ET (avg, 10x): ${avgTime} ms`;
    outputDiv.appendChild(avgTimeDiv);

    let avgMemoryDiv = document.createElement("div");
    avgMemoryDiv.textContent = `RAM (avg, 10x): ${avgMemory} MB`;
    outputDiv.appendChild(avgMemoryDiv);
}

function runJSBenchmark() {
    clearCell("javascript-output");
    n_digits_pi(10, 100_000); // Ejecutar con 100,000 dígitos
}

function clearCell(elementId) {
    document.getElementById(elementId).innerHTML = "";
}

File: 4.1-calculos-matematicos/Prueba-3/Variacion-2/python/main.py
import time
import tracemalloc
import js  # type: ignore
from pyscript import display  # type: ignore
from decimal import Decimal, getcontext


def calculate_pi_gauss_legendre(digits):
    # Configurar precisión arbitraria
    getcontext().prec = digits + 5  # Margen extra para evitar errores

    # Inicialización de variables
    a = Decimal(1)
    b = Decimal(1) / Decimal(2).sqrt()  # √2 con Decimal
    t = Decimal(1) / Decimal(4)
    p = Decimal(1)

    # ≈ log10(digits) iteraciones
    num_iter = int(Decimal(digits).log10().to_integral_value())

    for _ in range(num_iter):
        a_next = (a + b) / 2
        b = (a * b).sqrt()
        t -= p * (a - a_next) ** 2
        a = a_next
        p *= 2

    pi = ((a + b) ** 2) / (4 * t)
    return pi


def n_digits_pi(repetitions, digits):
    total_time = 0
    total_memory = 0

    start_total = time.time()

    for _ in range(repetitions):
        tracemalloc.start()

        start = time.time()
        pi_value = calculate_pi_gauss_legendre(digits)
        end = time.time()

        memory_usage = tracemalloc.get_traced_memory()[1] / (1024 * 1024)  # MB
        tracemalloc.stop()

        total_time += (end - start) * 1000
        total_memory += memory_usage

    end_total = time.time()
    total_exec_time = round((end_total - start_total) * 1000, 2)

    avg_time = round(total_time / repetitions, 2)
    avg_memory = round(total_memory / repetitions, 2)

    display(f"Total ET (10x): {total_exec_time} ms",
            target="pyscript-output")
    display(f"ET (avg, 10x): {avg_time} ms", target="pyscript-output")
    display(f"RAM (avg, 10x): {avg_memory} MB", target="pyscript-output")


def run_py_benchmark(event):
    js.clearCell('pyscript-output')
    n_digits_pi(10, 100_000)  # Se ejecuta con 100,000 dígitos
